### PR TITLE
feat: added ability to approve the service instead of the token manager

### PR DESCRIPTION
**Pull request by Foivos**

### CODE REVIEW

File: contracts/token-manager/TokenManagerLockUnlock.sol [hunk 1]
```diff
@@ -56,7 +56,10 @@ contract TokenManagerLockUnlock is TokenManager, ITokenManagerLockUnlock {
     function _takeToken(address from, uint256 amount) internal override returns (uint256) {
         IERC20 token = IERC20(this.tokenAddress());
 
-        token.safeTransferFrom(from, address(this), amount);
+        // slither-disable-next-line var-read-using-this
+        try interchainTokenService.transferFromSenderToTokenManager(this.interchainTokenId(), address(token), from, amount) {} catch {
```
[the remark this answers is not in the document]
Comment (Foivos) [reply contracts/token-manager/TokenManagerLockUnlock.sol:60]: if we enforce the pattern there is a better way of doing this, and we make the service also mint burn (we just delegate call to the implementations, which either have the service trasnferFrom or burn/burnFrom)

File: contracts/token-manager/TokenManagerLockUnlockFee.sol [hunk 2]
```diff
@@ -57,7 +57,11 @@ contract TokenManagerLockUnlockFee is TokenManager, ReentrancyGuard, ITokenManag
         IERC20 token = IERC20(this.tokenAddress());
         uint256 balanceBefore = token.balanceOf(address(this));
 
-        token.safeTransferFrom(from, address(this), amount);
+        if(token.allowance(from, address(this)) < amount) {
```
Comment (milapsheth) [inline contracts/token-manager/TokenManagerLockUnlockFee.sol:60]:
```suggestion
        if (token.allowance(from, address(this)) < amount) {
```

File: contracts/interfaces/IInterchainTokenService.sol [hunk 3]
```diff
@@ -188,6 +188,15 @@ interface IInterchainTokenService is
...
     ) external payable;
 
+    /**
+     * @notice Can be called by lock unlock token managers to allow for approvals to the service instaed of them.
+     * @param tokenId The unique identifier of the token to be transferred.
+     * @param token The address of the token to transfer from.
+     * @param from The address of the user to take tokens from.
+     * @param amount The amount of tokens to be transferred.
+     */
+    function transferFromSenderToTokenManager(bytes32 tokenId, address token, address from, uint256 amount) external;
```
Comment (milapsheth) [inline contracts/interfaces/IInterchainTokenService.sol:198]:
```suggestion
    function transferToTokenManager(bytes32 tokenId, address token, address from, uint256 amount) external;
```
a bit simpler perhaps?

**Review by milapsheth — APPROVED** [review]

**Review by deanamiel — APPROVED** [review]

**Comment by milapsheth** [comment]
Need to handle MintBurnFrom

**Pull request merged**